docs/src/service-workers.md

```diff
@@ -0,0 +1,349 @@
+await page.GotoAsync('/example-with-a-service-worker.html');
+var serviceworker = await waitForServiceWorkerTask;
+```
+
+[`event: BrowserContext.serviceWorker`] is fired ***before*** the Service Worker's main script has been evaluated, so ***before*** calling service[`method: Worker.evaluate`] you should wait on its activation:
+
+```js tab=js-ts
+await expect.poll(() => serviceworker.evaluate(() => (self as any).registration.active?.state)).toBe('activated');
+```
+
+```js tab=js-js
+await expect.poll(() => serviceworker.evaluate(() => self.registration.active?.state)).toBe('activated');
+```
+
+### Network Events and Routing
+
+Any network request made by the **Service Worker** will have:
+
+* [`event: BrowserContext.request`] and its correponding events ([`event: BrowserContext.requestFinished`] and [`event: BrowserContext.response`], or [`event: BrowserContext.requestFailed`])
+* [`method: BrowserContext.route`] will see the request
+* [`method: Request.serviceWorker`] will be set to the Service [Worker] instance, and [`method: Request.frame`] will **throw**
+* [`method: Response.fromServiceWorker`] will return `false`
+
+Additionally, any network request made by the **Page** (including its sub-[Frame]s) will have:
+
+* [`event: BrowserContext.request`] and its correponding events ([`event: BrowserContext.requestFinished`] and [`event: BrowserContext.response`], or [`event: BrowserContext.requestFailed`])
+* [`event: Page.request`] and its correponding events ([`event: Page.requestFinished`] and [`event: Page.response`], or [`event: Page.requestFailed`])
+* [`method: Page.route`] and [`method: Page.route`] will **not** see the request (if a Service Worker's fetch handler was registered)
+* [`method: Request.serviceWorker`] will be set to `null`, and [`method: Request.frame`] will return the [Frame]
+* [`method: Response.fromServiceWorker`] will return `true` (if a Service Worker's fetch handler was registered)
+
+Many Service Worker implementations simply execute the request from the page (possibly with some custom caching/offline logic omitted for simplicity):
+
+```js
+// filename: transparent-service-worker.js
+self.addEventListener("fetch", (event) => {
+  // actually make the request
+  const responsePromise = fetch(event.request); 
+  // send it back to the page
+  event.respondWith(responsePromise);
+});
+
+self.addEventListener("activate", (event) => {
+  event.waitUntil(clients.claim());
+});
+```
+
+If a page registers the above Service Worker:
+
+```html
+<!-- filename: index.html -->
+<script>
+  window.registrationPromise = navigator.serviceWorker.register('/transparent-service-worker.js');
+</script>
+```
+
+On the first visit to the page via [`method: Page.goto`], the following Request/Response events would be emitted (along with the corresponding network lifecycle events):
+
+| Event                             | Owner            | URL                            | Routed | [`method: Response.fromServiceWorker`] |
```

Comment:
Looks like this:

<img width="953" alt="Screen Shot 2022-07-13 at 2 20 51 PM" src="https://user-images.githubusercontent.com/11915034/178843237-2e91d43c-af9b-43e4-b005-1f18962ace37.png">